File: tests/cxx/java_like/Stack.cpp
#include "Stack.h"

int Stack::Node::getSum() const
//@ requires nodes(this, ?v, ?n, ?count, ?sum);
//@ ensures nodes(this, v, n, count, sum) &*& result == sum;
{
    int result = value;
    if (next != 0) {
        int next_sum = next->getSum();
        result += next_sum;
    }
    return result;
}

bool Stack::isEmpty() const
//@ requires StackPred(this, ?count, ?sum);
//@ ensures StackPred(this, count, sum) &*& result == (count == 0);
{
    //@ open StackPred(this, count, sum);
    return head == 0;
    //@ close StackPred(this, count, sum);
}

int Stack::peek() const
//@ requires StackPred(this, ?count, ?sum) &*& count > 0;
//@ ensures StackPred(this, count, sum);
{
    //@ open StackPred(this, count, sum);
    return head->value;
    //@ close StackPred(this, count, sum);
}

int Stack::getSum() const
//@ requires StackPred(this, ?count, ?sum);
//@ ensures StackPred(this, count, sum) &*& result == sum;
{
    int result;
    //@ open StackPred(this, count, sum);
    if (head != 0) {
        result = head->getSum();
    } else {
        result = 0;
    }
    //@ close StackPred(this, count, sum);
    return result;
}

void Stack::push(int value)
//@ requires StackPred(this, ?count, ?sum);
//@ ensures StackPred(this, count + 1, sum + value);
{
    //@ open StackPred(this, count, sum);
    Node *n = new Node(value);
    n->next = head;
    head = n;
    //@ close StackPred(this, count + 1, sum + value);
}

int Stack::pop()
//@ requires StackPred(this, ?count, ?sum) &*& count > 0;
//@ ensures StackPred(this, count - 1, ?nsum) &*& result == sum - nsum; 
{
    //@ open StackPred(this, count, sum);
    struct Stack::Node *head = this->head;
    int result = head->value;
    this->head = head->next;
    head->next = 0;
    delete head;
    //@ close StackPred(this, count - 1, sum - result);
    return result;
}

File: rust_tutorials/purely_unsafe/solutions/stack.rs
// verifast_options{ignore_unwind_paths}
use std::alloc::{Layout, alloc, handle_alloc_error, dealloc};
//@ use std::alloc::{Layout, alloc_block};

struct Node {
    next: *mut Node,
    value: i32,
}

struct Stack {
    head: *mut Node,
}

/*@

pred Nodes(node: *mut Node, count: i32) =
    if node == 0 {
        count == 0
    } else {
        0 < count &*&
        (*node).next |-> ?next &*&
        (*node).value |-> ?value &*&
        alloc_block_Node(node) &*&
        Nodes(next, count - 1)
    };

pred Stack(stack: *mut Stack, count: i32) =
    (*stack).head |-> ?head &*&
    alloc_block_Stack(stack) &*&
    0 <= count &*&
    Nodes(head, count);

@*/

impl Stack {

    unsafe fn create() -> *mut Stack
    //@ req true;
    //@ ens Stack(result, 0);
    {
        let stack = alloc(Layout::new::<Stack>()) as *mut Stack;
        if stack.is_null() {
            handle_alloc_error(Layout::new::<Stack>());
        }
        (*stack).head = std::ptr::null_mut();
        //@ close Nodes(0, 0);
        //@ close Stack(stack, 0);
        stack
    }

    unsafe fn push(stack: *mut Stack, value: i32)
    //@ req Stack(stack, ?count);
    //@ ens Stack(stack, count + 1);
    {
        //@ open Stack(stack, count);
        let n = alloc(Layout::new::<Node>()) as *mut Node;
        if n.is_null() {
            handle_alloc_error(Layout::new::<Node>());
        }
        (*n).next = (*stack).head;
        (*n).value = value;
        (*stack).head = n;
        //@ close Nodes(n, count + 1);
        //@ close Stack(stack, count + 1);
    }

    unsafe fn pop(stack: *mut Stack) -> i32
    //@ req Stack(stack, ?count) &*& 0 < count;
    //@ ens Stack(stack, count - 1);
    {
        //@ open Stack(stack, count);
        let head = (*stack).head;
        //@ open Nodes(head, count);
        let result = (*head).value;
        (*stack).head = (*head).next;
        dealloc(head as *mut u8, Layout::new::<Node>());
        //@ close Stack(stack, count - 1);
        result
    }

    unsafe fn dispose(stack: *mut Stack)
    //@ req Stack(stack, 0);
    //@ ens true;
    {
        //@ open Stack(stack, 0);
        //@ open Nodes(_, _);
        dealloc(stack as *mut u8, Layout::new::<Stack>());
    }

}

fn main()
//@ req true;
//@ ens true;
{
    unsafe {
        let s = Stack::create();
        Stack::push(s, 10);
        Stack::push(s, 20);
        Stack::pop(s);
        Stack::pop(s);
        Stack::dispose(s);
    }
}
